### ISSUE TITLE
Discuss features

### ISSUE THREAD

**Issue by kevinelliott** [shardbox/shardbox-web#1]
We need to discuss features (initial/MVP, then V1)

**Comment by vladfaust** [comment]
Users should be able to propose shards on their own, and the must be an admin interface to moderate entries. The core features should be categorizing and display dependencies (similar to https://shards.info). That's enough for MVP. I'm not sure if shardbox should be a shard repository, though.

**Comment by straight-shoota** [comment]
Management of entries all happens in https://github.com/shardbox/catalog and a PR-based editorial workflow should work well. Everyone can contribute and shard authors usually know their way around. Github is our admin interface ;)

Ruby toolbox uses the same approach: https://github.com/rubytoolbox/catalog

**Comment by straight-shoota** [comment]
Synchronization to the database is already implemented in [`import_catalog`](https://github.com/shardbox/shardbox-core/blob/master/src/service/import_catalog.cr). It's not complete yet, there are still some rough and untested edges with actions like removing and renaming categories. Turns out the implementation of these things is not as trivial as it sounds at first 😆 But it's already working for the most part and I'm sure this is way better than running our own managment interface with user authorization, review system etc.

**Comment by kevinelliott** [comment]
I do think we will outgrow a GitHub based catalog. I think it might be fine for the shard base data (i.e. the existence of the shard), but we will certainly need content that will not live there.

**Comment by kevinelliott** [comment]
The benefit of managing this in the database directly is that you can tie together various actions and intention around changes to the shard db. With changes to the yaml, you have little clue on what the intention is... was the shard renamed? was the shard removed? etc.

**Comment by straight-shoota** [comment]
> certainly need content that will not live there

What kind of user-created content do you have in mind?

**Comment by kevinelliott** [comment]
So if a shard author wants to publish, update, remove, rename, etc, we know the intention (either via a web action, or via a CLI tool we can make to help shard authors control that intention).

**Comment by kevinelliott** [comment]
This also allows us to grow into being a repository for shards as well, should we choose to do such a thing -- not in the sense of hosting the files (they are on GitHub, etc) -- but in the sense of allowing shard authors to publish, etc.

**Comment by kevinelliott** [comment]
The simplicity of using git to manage the catalog is alluring -- for sure it's simple, it's inspectable, it's safe, etc. It's also very limiting. So it's good today, but tomorrow it may hinder the ability to add flexibility and functionality around the management of shards.

**Comment by vladfaust** [comment]
I'm in for staying with git-based approach for now. @straight-shoota does it support a single shard having multiple categories?

**Comment by kevinelliott** [comment]
I imagine a world where authors can do:

```
shardbox create myshard   (wraps shards command to initialize, and preps for shardbox)
shards build
shardbox publish
```

As you can see, this would be inhibited by a GitHub-centric shard catalog.

**Comment by straight-shoota** [comment]
> So it's good today, but tomorrow it may hinder the ability to add flexibility and functionality around the management of shards.

For the current scope of the application it's solid and works. Should come the day that we need something else, we can change to a different workflow. There cost of change is not very high. But for it's the easiest and most efficient way.

@kevinelliott That `shardbox publish` etc. could work, even if it just creates a PR to add the shard to the respective category yaml 🤷‍♂

**Comment by straight-shoota** [comment]
> does it support a single shard having multiple categories?

Of course.

**Comment by kevinelliott** [comment]
I'm fine with that being for MVP, but now that we're all working on it, I'd like to see us really think it through for V1. It's still very limiting IMHO, and whatever we release with now is what the community would start to get use to. Very difficult to change these sorts of things later (have to educate people around the new flows, etc).

**Comment by kevinelliott** [comment]
The shards end up in the database anyway, and as you mentioned, what do you do in the case of renames, retraction/deletion, etc.

**Comment by kevinelliott** [comment]
Another option - we could reclassify the catalog as hints, and use it to seed the shard database (as well as look for additions and minor changes), still allow direct entry (via CLI/API, web forms, etc), but not use it as the source of truth for all things. Though I think this complicates things unnecessarily.

**Comment by vladfaust** [comment]
[KISS](https://en.wikipedia.org/wiki/KISS_principle)

**Comment by kevinelliott** [comment]
Of course. But, this is way too simple.

**Comment by kevinelliott** [comment]
🤷‍♂

**Comment by straight-shoota** [comment]
The catalog is already not the single source of shards, because when a shard is analyzed, its dependencies are also referenced to the database and a new entry inserted if a dependency doesn't exist yet. But that's not an issue. These are simply uncategorized shards. Renaming a shard should be really rare, and I'm not even sure how we should handle that at all. I might even go with treating a renamed shard as a completely separate entity.

I don't think this approach is to minimalistic. For ruby toolbox it seems to work great and that's almost exactly our use case right now.

**Comment by m-o-e** [comment]
On a somewhat meta-note:

I believe it is important that shardbox.org clearly communicates
that it is not an exhaustive catalog of available shards.

The headline should be changed from `Shards Database`
to something like `Manually curated catalog of popular shards`.

Ideally it should also link to the more complete catalogs
(at least https://crystalshards.xyz/) directly under the headline.

In its current form it could badly mislead especially newcomers to the
crystal language about the size of the ecosystem, as it covers 
only 10% of the available shards.

**Comment by straight-shoota** [comment]
> it is not an exhaustive catalog of available shards.

Maybe it's not completely there yet. But it's definitely meant to be that. And it certainly doesn't claim it would be exhaustive.

> The headline should be changed [...] to something like `Manually curated catalog of popular shards.`

That completely misses the scope of this project. Shardbox.org is *not* manually curated. The [base catalog](https://github.com/shardbox/catalog) is, but other shards are discovered automatically when used as dependencies. And it is not limited to popular shards, but every Crystal shard that's of any usefulness (see [guidelines](https://github.com/shardbox/catalog#guidelines)).

> Ideally it should also link to the more complete catalogs (at least https://crystalshards.xyz/) directly under the headline.

Describing crystalshards as "more complete" is disputable. It lists more entries than shardbox, but while shardbox lists shards, crystalshards lists only individual repositories. Many of these repos are not really useful as shards. And crystalshards completely misses out on shards hosted elsewhere than GitHub.
While in general, recall on crystalshards might still be higher, precision is definitely not (see next point).

Still, adding a link to crystalshards might be worth considering. But it should rather go into a *see also* section.

> In its current form it could badly mislead especially newcomers to the
> crystal language about the size of the ecosystem, as it covers
> only 10% of the available shards.

Actually, I believe the very opposite is true. crystalshards suggests there are thousands of shards available. In reality, the ecosystem is not that big. Crystalshards simply lists GitHub repositories tagged as `language:Crystal`. That does not even remotely match the definition of shards. It's more like a search for Crystal projects. There are solutions for advent-of-code, empty repositories, unfinished WIPs, demo prototypes, etc.

If you want to search broad, crystalshards is still the better tool. But shardbox is going to catch up quantitatively, and IMO is already superior qualitatively.

**Comment by m-o-e** [comment]
> doesn't claim it would be exhaustive.

Well, the headline `Shards Database` suggests exhaustiveness to me.
(but admittedly the other shard search sites don't describe their method
very visibly either)

> every Crystal shard that's of any usefulness

At the moment it unfortunately doesn't.
The following keywords miss many shards that I've personally
used: protobuf, cache. (I imagine it looks similar for other keywords)

> Crystalshards simply lists GitHub repositories tagged

Hm, interesting, I never noticed that. 🤔 
(it has never led me to a repo that wasn't a shard)

Well, in any case, I like the idea of shardbox.

My only concern is really that new users could mistake it
for an exhaustive search and - in the worst case - abort their
evaluation of crystal when they are led to believe that important
baseline shards (like protobuf) don't exist.

**Comment by straight-shoota** [comment]
Yeah the title was just born as a filler and never really put any thought into it. The HTML page title is more explicit: `Shardbox: A shards database`. So maybe we could just add an indefinite article to the on-page headline to match that?

> when they are led to believe that important
baseline shards (like protobuf) don't exist.

The best solution to that problem would be to add missing shards to the catalog. 😄

**Comment by m-o-e** [comment]
> The best solution to that problem would be to add missing shards to the catalog.

Well, I'm afraid most shard authors don't know that shardbox even exists
and many may not have time or interest to go through the effort of
figuring out the PR process to have their shard added. :(

**Comment by straight-shoota** [comment]
Everyone can add shards to the catalog, not just the authors. In fact, it's relatively easy to add them in bulk. I have been focusing on ironing out features and fixing bugs. So any help with expanding the catalog is welcome.

**Comment by chances** [comment]
Couldn't some of this manual catalog work be automated?

When Shardbox discovers a new shard it can use the shard's `description` to add a new catalog entry.

Also consider an unseen GitHub repo that has Topics (e.g. [game-engine](https://github.com/topics/game-engine)) set by the repo owner. Shardbox could use that to place shards in appropriate catagories.

**Comment by straight-shoota** [comment]
That could be an option, yes.
But I think there should still be some kind of review process, so someone would need to take a look at these automatic catalog entries. At that point I'm not sure if it makes much of a difference whether the reviewer only reviews automatically created entries or initiates categorization directly.
Both workflows need tools to function efficiently. And I think a tool that helps human editors is more versatile. It could be used by authors to add their shards to the catalog and it can be used by catalog contributers to edit in bulk.
I actually have a basic tool for that working. It's not pretty and currently only works locally with access to the git repo and database server. But it's really great for categorizing lots of shards.
